pkg/instrumentation/custom/collector.go

```diff
@@ -0,0 +1,76 @@
+package instrumentation
+
+import (
+	"context"
+	"time"
+
+	"github.com/content-services/content-sources-backend/pkg/dao"
+	"github.com/content-services/content-sources-backend/pkg/instrumentation"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/rs/zerolog/log"
+	"gorm.io/gorm"
+)
+
+const tickerDelay = 5 // in seconds // could be good to match this with the scrapper frequency
+
+type Collector struct {
+	context context.Context
+	metrics *instrumentation.Metrics
+	dao     dao.MetricsDao
+}
+
+func NewCollector(context context.Context, metrics *instrumentation.Metrics, db *gorm.DB) *Collector {
+	if context == nil {
+		return nil
+	}
+	if metrics == nil {
+		return nil
+	}
+	if db == nil {
+		return nil
+	}
+	return &Collector{
+		context: context,
+		metrics: metrics,
+		dao:     dao.GetMetricsDao(db),
+	}
+}
+
+func (c *Collector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.metrics.RepositoriesTotal.Desc()
+	ch <- c.metrics.RepositoryConfigsTotal.Desc()
+	ch <- c.metrics.PublicRepositoriesNotIntrospectedLast24HoursTotal.Desc()
+	ch <- c.metrics.PublicRepositoriesWithFailedIntrospectionTotal.Desc()
+	ch <- c.metrics.NonPublicRepositoriesNotIntrospectedLast24HoursTotal.Desc()
+}
+
+func (c *Collector) Collect(ch chan<- prometheus.Metric) {
```

Comment:
yes and no! I wrongly added the Collect interface, but it is not being used, because the code that register this is commented;  so I thought I was using that (but not used) and I thought I was gathering the values using that interface (but it is not the case); I have removed as that could evoke wrong behaviors. By the way, I think I have implemented wrongly.

What I think about when it should be used that interface:
- When I am defining a new more complex new metric not covered by the basic types, so one method gather the description of the metric, and the other write all the values of the metric (it could be several values generated as an histogram, or several values with different labels, etc...); I name it value to distinguish from the metric that could have several values; a gauge with labels could print out several values too and I give the scope of metric to one description.

What I think the dead code would do (wrongly):
- It would mix all the descriptions of the different metrics.
- It would write all the values together as part of the last description.
- It would not match the metrics descriptions with the expected metric value.

I wrongly assumed it was working because I saw the metrics into the curl request! many thanks mate!

Reply:
Ah okay, thanks! I was just wondering what the methods were for, but I'm glad I also accidentally helped you find a bug :).